### ISSUE TITLE
Can org-brain-visualize work like minimap or like speedbar, attached if the user wants, from a customizable variable?

### ISSUE THREAD

**Issue by techapu** [Kungsgeten/org-brain#165]
I want to tell you that the helm-brain "visualize" function works incredible well.
When I am in a candidate, if I press (C-j) it shows the entry without leaving helm. And it's bloody fast. 
So I can wonder about many visualize maps in many diferent places of the brain in seconds. 
It's so much better that the performance of the org-brain-visualize or the org-brain-goto-entry function that I am in the duty of telling you that this way is much more practical in terms of performance.
And it makes me think of a custom attached buffer posibility (attached to the org-mode buffer) with the org-brain-visualize opened, visiting the heading I'm in when I'm hoovering over it. When one needs to switch to that buffer, calling `org-brain-visualize-entry-at-pt` from a confortable keybinding will do the job. And pressing "o" to open current entry takes the user back to the editing situation.
Thanks for the patience! Sorry for the multiple entries, but I'm a heavy user of your package...
:)

**Comment by Kungsgeten** [comment]
I don't quite understand the workflow you're describing. Isn't it possible to do what you describe already, by not closing the `*org-brain*` buffer?

**Comment by techapu** [comment]
No, I mean to actually achieve all that operations from within helm-brain, without exiting. 
Anyway, from org-brain buffer it keeps opening a lot of new windows, one each time the entry belongs to a new file. And I suggested a buffer attached to the main org-mode one, like in the `interleave-mode` package, or an IDE environment. Like some kind of file tree, fixed lateral. And I suggest this as an optional customizable configuration, not the as the main one, that is very clever too. 
I am in Debian 10 Buster, with Emacs 26.1, using the last version of your package from Melpa. I have an i5 Kaby lake processor, 16gb RAM, a 256 SSD Kingston, and a Nvidia gtx1080ti card. I think I have no performance issues with the hardware. 
And the time from when I begin the task calling `org-brain-visualize` or `org-brain-goto-entry` is very very slow until it shows me the "Pattern:" prompt. I mean between thirty second up to two minutes. 
Even when I call `helm-brain` it can take between a minute or two, so I don't wanna exit after that. :)
But `helm-org-rifle-brain` takes like between 5 to 15 seconds. So that's why I insist with that package even when I understand it is an external package, because it makes your package performs a lot faster.
So reducing the time it takes to work, is very important for me from an user perspective. 

When we tried the functions initially coded by you in the Issue [#161](url), they worked very fast. But when you published the `helm-brain` function now it takes a lot more to initialize each time I call the function. I can't figure why, but I don't master the inner mechanics of your package.

Isn't there any caching method to reduce the lag?

**Comment by Kungsgeten** [comment]
My guess is that `helm-org-rifle` uses an external program like `ripgrep` or similar to search the text. `org-brain` use emacs lisp and no caching. Perhaps I should do some experimentation to use `ripgrep` inside of `org-brain` to increase performance. Running `org-brain-visualize` for me takes about 2 seconds on Windows (I think its faster on Linux) with 266 entries. The lag isn't really caused by `org-brain-visualize` but by the function `org-brain-choose-entries`.

Regarding a "permanent" buffer it could be nice to support something like [Treemacs](https://github.com/Alexander-Miller/treemacs). The downside would be that its a tree (only one parent) while `org-brain` is more like a graph. The upside is that its probably a lot easier to implement than a completely new sidebar feature.

**Comment by Kungsgeten** [comment]
@techapu I've been experimenting with a different branch of `org-brain`: https://github.com/Kungsgeten/org-brain/tree/brain2

Would you mind testing it? Please backup your current brain-folder first though. This branch only works on headline entries, so here the concept of "file entries" is removed. What is new though is that I've tried to implement entry searching with `ripgrep`, which should be faster.

If you try the branch, and have `ripgrep` installed, use `M-x org-brain-setup-rg-bindings`, which will replace the bindings in `org-brain-visualize-mode`. You can also use the commands starting with `org-brain-rg` using `M-x`.

I would be happy if you can report the difference in speed on your computer with this branch, comparing `org-brain-visualize` and `org-brain-rg-visualize`.

**Comment by MarioRicalde** [comment]
@techapu are you still interested in pursuing this? @Kungsgeten prepared something for you since June 16, 2019.

**Comment by Kungsgeten** [comment]
The branch I linked to is irrelevant. I've decided to continue to support both headline and file entries.

**Comment by Kungsgeten** [comment]
@techapu The latest version of `org-brain-visualize` should be much faster than before (although it can be slow the first time it is run after an Emacs reboot). Please try it out in your config and see how it compares to `helm-brain`.
